Exclude join if needed

falkodev opened the pull request:
In previous PRs (https://github.com/apostrophecms/apostrophe-headless/pull/23 and https://github.com/apostrophecms/apostrophe-headless/pull/25), I added an option `api` to fields to hide them in GET requests.

But I did not check joins, and I realized it was not working (so joins were still sent in the response with `api: false`or `api: 'editPermissionRequired'` in the schema). I think this is because joins are fetched when the cursor is called.

Here is my proposal to solve this (and even to be able to hide only a relation in a join field if needed). The downside of this solution is this is done after the `findForRestApi` function.

boutell commented:
You can chain `.joins(false)` on the cursor, or `.joins([ array of join names... ])`. That would yield a performance win too by actually skipping the queries, vs. just dropping the data after doing the hard work.

falkodev commented:
@boutell 
Thank you for your comment. It is much easier this way 😃 

But I also needed to remove the relationship. By using only the `joins` option in the cursor, there were still the `idsField` and `relationshipsField` visible with the data I wanted to hide in the first place.

So, I modified the function `removeKey` to take into account these additional fields as well. By doing this, I realized I did not need the `joins` option anymore. I believe it is because the join is fetched when there is a `idsField` (or `idField`).

boutell commented:
This is going in a great direction, unfortunately you can't mix `0` and `1` in a projection. Right now if `includeFromQuery` is true, the projection might contain both `1`'s and `0`'s (from `removeJoin`). For this reason in this situation you will need to use the `joins` filter. You would need to iterate the fields to figure out what joins *should* be included, and pass an array of those names to the `joins` filter.

At that point, filtering out the idField and idsField from the projection becomes optional.

falkodev commented:
Maybe my way of doing it is not the best. But, first there is this code:
```javascript
        if (includeFromQuery || projection.hasOwnProperty(field.name)) {
          delete projection[field.name];
          if (field.type.match(/join/)) {
            joinsToRemove.push(field);
          }
        }
```

and later: 
```javascript
      if (_.isEmpty(projection) && joinsToRemove.length > 0) {
        joinsToRemove.forEach(function(join) {
          removeJoin(join);
        })
      }
```

So, I add `0`s only when there is no `1`s. I also added this test:
```javascript
it('can GET a product without excluded joins even if included in query', function(done) {
    return http('/api/v1/products?includeFields=slug,type,_articles', 'GET', {}, {}, undefined, function(err, response) {
      assert(!err);
      assert(response);
      assert(response.results);
      var product = _.find(response.results, { slug: 'product-key-product-with-join' });
      assert(typeof product.type === 'string');
      assert(typeof product.slug === 'string');
      assert(typeof product['_articles'] === 'undefined');
      done();
    });
  }); 
```

But if you find it more readable to use `joins`, I can do this:
```javascript
    self.schema.forEach(function(field) {
        if (field.api === false) {
          removeKey(field);
        }
        if (field.type.match(/join/)) {
          joins.push(field.name);
        }
      })

...

// in function removeKey
     if (includeFromQuery || projection.hasOwnProperty(field.name)) {
          delete projection[field.name];
          if (field.type.match(/join/)) {
            joinsToExclude.push(field.name);
          }
        }

...

      var joinsToInclude = _.difference(joins, joinsToExclude);
      var cursor = self.find(req, {})
        .projection(projection)
        .joins(joinsToInclude)
        .safeFilters(options.restApi.safeFilters || [])
        .queryToFilters(req.query, which);
```

So, `removeJoin` is not called anymore with `includeFromQuery` and the result is the same (my manual tests, and the unit tests I added both work ok with all sort of `includeFields` and `excludeFields`)

Let me know what is the best fit.

boutell commented:
Looks good, thank you for the thorough testing.

boutell commented:
Published as 2.6.0. Thanks again.
